### ISSUE TITLE
Allow 2 config files (Standalone and VST plugin)

### ISSUE THREAD

**Issue by elieserdejesus** [elieserdejesus/JamTaba#104]
Store the user configs in separated files for standalone and VST plugin

**Comment by Ezeer** [comment]
My idea is to create a folder inside Jamtaba 2 for the plugin .
You don't have to change the name of the file , just make sure he will not be overwritten .
Now i think that this should be in the folder of the vst plugin .

**Comment by jonjamcam** [comment]
I think it's more hidden where it is now, but you can name it jamtaba2v,json. no need for folder. what do you thnk

**Comment by Ezeer** [comment]
**I take an exemple with DSK :** 

![dsk](https://cloud.githubusercontent.com/assets/1283808/11026446/5272c378-86aa-11e5-92af-efac1d1b1053.jpg)

Look to the full path , then the folder content and the subFolder .
If you think it must be hidden , put it there ( subFolder, DSK use their data like this ) .
But the .json and logging.ini must be tweaked some times , we don't want hide it too much .

**My simulation now** : 

![vst](https://cloud.githubusercontent.com/assets/1283808/11026577/5cbec028-86ac-11e5-8a96-68c47ddfa01b.jpg)


In Data there is : 

![vst3](https://cloud.githubusercontent.com/assets/1283808/11026570/405db10a-86ac-11e5-934c-63ee35de695d.jpg)

**Comment by Ezeer** [comment]
A variant : 

![vst4](https://cloud.githubusercontent.com/assets/1283808/11026626/1752eafe-86ad-11e5-8534-c39d783831b9.jpg)
![vst5](https://cloud.githubusercontent.com/assets/1283808/11026629/1d8381f4-86ad-11e5-9293-e87b3e4cde70.jpg)

**Comment by jonjamcam** [comment]
i see. Probably is better to create a Jamtaba folder and there put everything. And in case someone deletes the json file the plugin creates a new one when launched, doesn't it?

**Comment by Ezeer** [comment]
Yes it does .

**Comment by elieserdejesus** [comment]
Ezee, is better avoid write files in the Vst plugin directory because admin privileges.
For example, if users install the plugin in the Steinberg recommended folder (C:/Program Files/Steinberb/VstPlugins/Jamtaba) Jamtaba needs admin privileges to write the json in this folder every time user change something in your preferences.

The installer can write files in this folder because installers has God privileges.

The best approach is store data in AppData, in this folder we don't have problems with privileges.

**Comment by Ezeer** [comment]
copy that !
We keep that folder approach now  ? ( separated plugin and standalone , plugin being a sub folder
of Jamtaba 2 ? )

**Comment by elieserdejesus** [comment]
Ezee, I propose 2 folder, but share the cache.bin:
```
AppData/Jamtaba2
    cache.bin
    Standalone
         logging.ini
         log.txt
         jamtaba.json  
    VstPlugin
         logging.ini
         log.txt
         jamtaba.json  

```

**Comment by Ezeer** [comment]
You are the boss , boos !
( also you write the installer stuff eh eh )

My taste : 
![idea](https://cloud.githubusercontent.com/assets/1283808/11048706/d6338ce4-8739-11e5-9313-4cd2bc617867.jpg)

**Comment by Ezeer** [comment]
@jonjamcam Do you see waddah see ?

![idea2](https://cloud.githubusercontent.com/assets/1283808/11048935/03c9512e-873b-11e5-907c-251c40960791.jpg)

:+1:

**Comment by elieserdejesus** [comment]
Let's use this approach (a VstPlugin folrder inside Jamtaba 2 folder).

**Comment by Ezeer** [comment]
You are my prefered boss !
![boss-3v](https://cloud.githubusercontent.com/assets/1283808/11049255/6e2e9022-873d-11e5-8556-210b0bef1b0f.jpg)

**Comment by elieserdejesus** [comment]
Ezee, the changes are in the code, not in installer. Are you interested?  :)

**Comment by Ezeer** [comment]
Of course i am , a hint ?

**Comment by elieserdejesus** [comment]
The path of log file is used in logging.cpp, line 60. The problem is we have no context to discover if we running as plugin or as standalone. The logging function is just a function at moment. Maybe you can create a new function in MainController, using the same signature of the logging function. If Qt accept a member function (Maincontroller::loggingFunction) this ideas works, because now the logging function has some context to discover if the mainController is running as vst plugin or not.

The logging function is installed in main.cpp, line 21
```
qInstallMessageHandler(jamtabaLogHandler);
```
For the other files we need think in more details like these...

**Comment by Ezeer** [comment]
##The problem is we have no context to discover if we running as plugin or as standalone. ##
By using the QApplication ?
>QApplication::setApplicationName("Jamtaba 2");

**Comment by Ezeer** [comment]
then : 
> QApplication::instance()->applicationVersion()

?

**Comment by elieserdejesus** [comment]
Ezee, I think the 
```
QApplication::setApplicationName("Jamtaba 2 StandAlone");
```
is a bit dangerous. The Qt StandardPaths use the QApplication::applicationName(), so if change the application name other paths will be broken.

Please create another issue to handle this showUp stuff in SingleApplication.

**Comment by Ezeer** [comment]
" is a bit dangerous. "
Less than me ( lol )
https://www.youtube.com/watch?v=aLJFRgE4Ywk

We shall use the power of Qt everywhere , so we got options : 
http://doc.qt.io/qt-5/qapplication-members.html
make your choice ...

mine could be : 

>QString	applicationDisplayName()
void	setApplicationDisplayName(const QString & name)

**Comment by elieserdejesus** [comment]
Nice Ezee, I think applicationDisplayName will work, please try it.

**Comment by Ezeer** [comment]
Ok , i'll try it so .
Thx elieser .
:+1:

**Comment by Ezeer** [comment]
@elieserdejesus  I've seen " The path of log file is used in logging.cpp, line 60 "
But i need to create a folder for the plugin . You say it's not the installer stuff ?

**Comment by elieserdejesus** [comment]
Yes Ezee, mkPath is creating the folder. To create a VSTPlugin folder you can use
```
logDir.mkpath("VSTPlugin");
```

**Comment by Ezeer** [comment]
Ok , now i've got a little problem : 
void setApplicationDisplayName(const QString & name) no more exists ...

So i think we should use QApplication::setApplicationVersion(APP_VERSION); in main.cpp
Are you ok with that ? ( to set get plugin or standalone version , sounds logical ... )

**Comment by elieserdejesus** [comment]
Ezee, setApplicationDisplayName is defined in [QGuiApplication](http://doc.qt.io/qt-5/qguiapplication.html#applicationDisplayName-prop)

and QApplication inherit from QGuiApplication.

We are using QApplication in SingleApplication stuff.

It's not just a question of some cast to QApplication?

APP_VERSION  is 2.0.5, and I'm planning use this to auto update. This number is sended to server. So is not a good idea use APP_VERSION to identify the plugin or standalone.

**Comment by Ezeer** [comment]
**It's not just a question of some cast to QApplication?**
woaow , i'm tired  ....
 QApplication::setApplicationDisplayName("Jamtaba Standalone"); works in fact , i have used it on main .

it's QString applicationDisplayName() that is unknown for Qt ...
But it's a minor problem , i will find a solution , thanks .

**Comment by elieserdejesus** [comment]
Ezee, to get the application display name you can use
```
QString appName = QApplication::applicationDisplayName();
```

**Comment by Ezeer** [comment]
Thanks .
but ...
I don't like how you designed the logging and the creation of Jamtaba2 folders .
That works , but that is not art ... I think that part should be improved , cleaned .

--------------------------------------------------------------------------------------------------
QString logFile = Controller::MainController::getLogConfigFilePath();
       if(!logFile.isEmpty()){
        qputenv("QT_LOGGING_CONF", QByteArray(logFile.toUtf8()));
        qInstallMessageHandler(jamtabaLogHandler); <------ weird inside ...
    }

    Persistence::Settings settings;
    settings.load(); <---------- You know that will fail at 1rst launch ....
----------------------------------------------------------------------------------------------------    
Perhaps the folder and Json should exist before settings.load() .
Also after closing Jtb for the first time , there is no log created .
Perhaps i'm too much tired , but that design is very weird and unfamiliar to me .
**QDir logDir(QStandardPaths::writableLocation(QStandardPaths::AppLocalDataLocation));**
weird to see that in jamtabaLogHandler ,  why not create it before ?   
I'll work on it tomorrow ...

**Comment by elieserdejesus** [comment]
Ezee, I agree about the weirdness, but some of this weird is related to Qt restrictions.

For example, 
```
qputenv("...");
qInstallMessageHandler( jamtabaLogHandler );
```
are Qt functions. The first is setting the log rules readed from logging.ini file in Qt environment.
The second (qInstallMessageHandler) is another Qt function. This function expect a callback function as parameter: the logHandler function. Of course the signature of this callback function is a bit weird, using some alien parameters, but we can't change this without break Qt logging system.

I agree about your comment in
```
Persistence::Settings settings;
settings.load(); <---------- You know that will fail at 1rst launch ....
```
but I have no idea about how to solve this in a more elegant way.

-----------------------------
My preferred idea about the logHandling is try use a member function, and remove the simple callback function.

Something like this:
```
class MainController{
      public:
           MainController(){
                QString logFile = getLogConfigFilePath();//now we know if we are running or not as a VST plugin
                if(!logFile.isEmpty()){
                       qputenv("QT_LOGGING_CONF", QByteArray(logFile.toUtf8()));
                       qInstallMessageHandler( logHandler ); ///using a member function as logHandler
                 }
           }

      private:
            void logHandler( weird but necessary qt parameters ){
             //loggind details here, and here we know if we are running or not as VST plugin
             }
};

**Comment by Ezeer** [comment]
I'm too much tired to take any decision , i just know that your Settings class is beautiful , it's a shame
to fail at startup ... lol .

Your idea seems to be nice , i'll check it better tomorrow with a brand new neuronal system .
Thx elieser .

**Comment by elieserdejesus** [comment]
No problem Ezee, take a good rest!

**Comment by elieserdejesus** [comment]
Ezee, I'm thinking here...
In my opinion we have to put this feature in the last item of our TODO list. Let's work in all other minor features for first, and work in this "2 folder" stuff when we are finishing the release. I propose this because I'm afraid this feature will have a huge impact in many places of the code. So, If you work on this feature, and I work in other features in parallel, we possibly are changing the same source files, and merge conflicts will be a pain.

**Comment by Ezeer** [comment]
Yes i am working on it , but i don't think that will impact to the global program .
I use a new class that do all the work that was dispatched into several class .
That class is external , initialized in main and destroyed here : 
void StandaloneMainController::quit()
{
    //destroy the extern !
    if(JTBConfig)delete JTBConfig;
    application->quit();
}

But it's okay , i can focus on other things if you prefer .
I've seen your performance counter problem for example , that is a priority now .

**Comment by elieserdejesus** [comment]
Ezee, if your solution will not impact in many sources keep working on this. 
The idea of a .h file different .cpp to standalone and vst plugin change a lot of things, but if you have another and simple solution keep working.

**Comment by Ezeer** [comment]
I've got good news .
The class Configurator is able to create the home directory tree , create the files and read the JSON
depending on how you setup it : 

int main(int argc, char* args[] ){

    QApplication::setApplicationName("Jamtaba 2");
    QApplication::setApplicationVersion(APP_VERSION);
   
    //start the configurator
    JTBConfig=new Configurator();
    if(!JTBConfig->setUp(plugin)) qWarning() << "JTBConfig->setUp() FAILED !" ;
     ...

It is very fun to use it as dev now, because i only have to change on parameter to access 2 different
setups ! JTBConfig->setUp(**plugin** or **standalone**) Magic ... That opens new horizons , like the ability to record different setups in the future .
I don't have finished so i won't publish the code now, and i can't code and test the plugin part .
So when i will have worked enough with it i will create a branch here to allow you to pull it and test
on the plugin side .

**Comment by Ezeer** [comment]
Added : **Q_DECLARE_LOGGING_CATEGORY(jtConfigurator)**
Result in console ( Qdebug ...)
>Starting E:\JamTaba\build-Jamtaba-Desktop_Qt_5_5_1_MinGW_32bit-Release\Standalone\release\Jamtaba2.exe...
jt.Configurator:  HomePath folder CREATED !
jt.Configurator:  Creating PluginVst folder...
jt.Configurator:  PluginVst folder CREATED !
jt.Configurator: Ini file copied in : "C:/Documents and Settings/Salon/Local Settings/Application Data/Jamtaba 2/logging.ini"

**Comment by Ezeer** [comment]
DOne for the standalone project .
I can't compile the plugin actually , so it's up to you to add the two files ( header+cpp)
in the plugin project and to implement it elieser .
Really easy , you'll see : https://github.com/elieserdejesus/JamTaba/tree/configurator
